libraries/chain/db_market.cpp

```diff
@@ -364,18 +364,137 @@ bool database::apply_order(const limit_order_object& new_order_object, bool allo
+      {
+         auto old_limit_itr = limit_itr;
+         ++limit_itr;
+         // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+         finished = ( match( new_order_object, *old_limit_itr, old_limit_itr->sell_price ) != 2 );
+      }
+
+      if( !finished )
+      {
+         // check if there are margin calls
+         const auto& call_price_idx = get_index_type<call_order_index>().indices().get<by_price>();
+         auto call_min = price::min( recv_asset_id, sell_asset_id );
+         auto call_itr = call_price_idx.lower_bound( call_min );
+         // feed protected https://github.com/cryptonomex/graphene/issues/436
+         auto call_end = call_price_idx.upper_bound( ~sell_abd->current_feed.settlement_price );
+         while( !finished && call_itr != call_end )
+         {
+            auto old_call_itr = call_itr;
+            ++call_itr; // would be safe, since we'll end the loop if a call order is partially matched
+            // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+            // assume hard fork core-338 and core-625 will take place at same time, not checking HARDFORK_CORE_338_TIME here.
+            finished = ( match( new_order_object, *old_call_itr, call_match_price ) != 2 );
+         }
+      }
+   }
+
+   // still need to check limit orders
+   while( !finished && limit_itr != limit_end )
+   {
+      auto old_limit_itr = limit_itr;
+      ++limit_itr;
+      // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+      finished = ( match( new_order_object, *old_limit_itr, old_limit_itr->sell_price ) != 2 );
+   }
+
+   const limit_order_object* updated_order_object = find< limit_order_object >( order_id );
+   if( updated_order_object == nullptr )
+      return true;
+
+   // before #555 we would have done maybe_cull_small_order() logic as a result of fill_order() being called by match() above
+   // however after #555 we need to get rid of small orders -- #555 hardfork defers logic that was done too eagerly before, and
+   // this is the point it's deferred to.
+   return maybe_cull_small_order( *this, *updated_order_object );
+}
+
 /**
- *  Matches the two orders,
+ *  Matches the two orders, the first parameter is taker, the second is maker.
  *
  *  @return a bit field indicating which orders were filled (and thus removed)
  *
  *  0 - no orders were matched
- *  1 - bid was filled
- *  2 - ask was filled
+ *  1 - taker was filled
+ *  2 - maker was filled
  *  3 - both were filled
  */
-template<typename OrderType>
-int database::match( const limit_order_object& usd, const OrderType& core, const price& match_price )
+int database::match( const limit_order_object& usd, const limit_order_object& core, const price& match_price )
 {
    assert( usd.sell_price.quote.asset_id == core.sell_price.base.asset_id );
```

Comment:
Actually we need to carefully review all `assert()`s and `FC_ASSERT()`s as mentioned in https://github.com/bitshares/bitshares-core/issues/679 and https://github.com/bitshares/bitshares-core/issues/511#issuecomment-366802721.

For this one, if it would cause a chain halt, most probably would be caused by FBA buyback.